CRM/Core/BAO/Address.php

```diff
@@ -362,127 +362,88 @@ public static function &getValues($entityBlock, $microformat = FALSE, $fieldName
+      if (isset($entityBlock[$fieldName])) {
+        $addressApi->addWhere($fieldName, '=', $entityBlock[$fieldName]);
+      }
     }
     else {
       $addressIds = [];
       $addressIds = self::allEntityAddress($entityBlock);
 
       if (!empty($addressIds[1])) {
-        $address->id = $addressIds[1];
+        $addressApi->addWhere('id', '=', $addressIds[1]);
       }
       else {
         return $addresses;
       }
     }
     if (isset($entityBlock['is_billing']) && $entityBlock['is_billing'] == 1) {
-      $address->orderBy('is_billing desc, id');
+      $addressApi->addOrderBy('is_billing', 'DESC');
     }
     else {
       //get primary address as a first block.
-      $address->orderBy('is_primary desc, id');
+      $addressApi->addOrderBy('is_primary', 'DESC');
     }
+    $addressApi->addOrderBy('id', 'ASC');
 
-    $address->find();
+    $addresses = (array) $addressApi->execute();
 
-    $locationTypes = CRM_Core_DAO_Address::buildOptions('location_type_id');
     $count = 1;
-    while ($address->fetch()) {
-      // deprecate reference.
-      if ($count > 1) {
-        foreach (['state', 'state_name', 'country', 'world_region'] as $fld) {
-          if (isset($address->$fld)) {
-            unset($address->$fld);
-          }
-        }
-      }
-
-      $values = [];
-      CRM_Core_DAO::storeValues($address, $values);
+    foreach ($addresses as $k => $address) {
+      $values = $address;
 
-      // add state and country information: CRM-369
-      if (!empty($address->location_type_id)) {
-        $values['location_type'] = $locationTypes[$address->location_type_id] ?? NULL;
-      }
-      if (!empty($address->state_province_id)) {
-        $address->state = CRM_Core_PseudoConstant::stateProvinceAbbreviation($address->state_province_id, FALSE);
-        $address->state_name = CRM_Core_PseudoConstant::stateProvince($address->state_province_id, FALSE);
-        $values['state_province_abbreviation'] = $address->state;
-        $values['state_province'] = $address->state_name;
-      }
-
-      if (!empty($address->country_id)) {
-        $address->country = CRM_Core_PseudoConstant::country($address->country_id);
-        $values['country'] = $address->country;
-
-        //get world region
-        $regionId = CRM_Core_DAO::getFieldValue('CRM_Core_DAO_Country', $address->country_id, 'region_id');
-        $values['world_region'] = CRM_Core_PseudoConstant::worldregion($regionId);
-      }
+      // add state and country information: CRM-369. Also necessary for vCard format.
+      // FIXME: Longer-term, CRM_Utils_Address::format and formatVCard should use API4 field names.
+      $values['location_type'] = $address['location_type_id:label'];
+      $values['state_province_abbreviation'] = $address['state_province_id:abbr'];
+      $values['state_province'] = $values['state_province_name'] = $address['state_province_id:label'];
+      $values['country'] = $address['country_id:label'];
```

Comment:
It probably makes sense to move `    $address['county'] = $address['county_id:label'] ?? NULL;` from `addDisplay` to here (maybe also the `address_id` line)- since then that fn just deals with the .... display